### DISCUSSION TITLE
H2 build report

### DISCUSSION THREAD

**Discussion by leehambley** [kachurovskiy/nanoels#220] [Show and tell]
I started on a H2 build more than a year ago, but due to incompatibility between my lathe housing, and the H2 enclosure I stalled for a while, also I rarely/never use my lathe and I didn't need to hurry. As a new owner of a Bambu X1, and having relearned FreeCAD after their 1.0 RC I decided to finish off the build.

The machine is a Optimum TU 2406 V - Vario; it's probably not dissimilar to any other asia-made lathe, however being a _vario_ (variable speed) it had some decent built-in electronics with a 5, and 9v supply on the power board.

I was able to mount the stepper motor below the lead screw, under the fascia using 1:1 ratio gears which I custom printed. I would like to change this for 1:3 pulleys so I get more torque, but the 3d printed gears (with a filed-in keyway) work basically fine for now, however they are noisy due to not running perfectly round.

I originally had a Honeywell DIN rail 5, and 24v power supply in an electrical enclosure below the lathe, however I couldn't source a single cabinet large enough to house the comically-large 24v din rail power supply. I had an inexpensive no-name 24v/3.5A DC supply which came with some CNC kit I never used hanging around. This 24v PSU fits inside the side cover where the belt-drive and is conveniently close to where the main power switch.etc exist.

For the 5v, the machine's built-in tachometer already had 5v, which I tapped and used to power the Arduino.

Attached a bunch of pics, I hope someone gets some benefit from this thread.

Edit to add: the freecad files in the h2 directory seem to be an older version than the stls in the h2 directory; I stumbled over this and modified the printed case directly with a finger sander.

Edit again to add: the hand-drawn labels are actually 3d-printed into the case, but I had such a hard time to model the icons and project the geometry to have neat icons, that I gave up and hand-drew them in the slicer.

![PXL_20241004_120535134](https://github.com/user-attachments/assets/75d5cf1c-3b6a-4343-a904-9cd69782d2d3)
![PXL_20241004_120632622](https://github.com/user-attachments/assets/a988632a-1e5f-4144-aaf9-e709cfc67d10)
![PXL_20241004_120547181](https://github.com/user-attachments/assets/00135b5d-4ce6-433e-a60e-8c50b99ab930)
![PXL_20241004_120639588](https://github.com/user-attachments/assets/d832fc9d-61ba-4bcc-affe-c28de2d0dbf8)
![PXL_20241004_120616541](https://github.com/user-attachments/assets/18b13cb5-6083-4e95-ab49-acaf7b19e9bc)
![PXL_20241004_120552879](https://github.com/user-attachments/assets/8ad45d4d-c534-47e1-a912-d80491f09813)
![PXL_20241004_120635995](https://github.com/user-attachments/assets/488e5cf7-1d98-459a-b62a-370162e712b3)
![PXL_20241004_120627318](https://github.com/user-attachments/assets/90f643ed-d945-4ad5-97d9-173e4880ca7a)
![PXL_20241004_120613750](https://github.com/user-attachments/assets/cdb37cc9-1ad7-49b1-b943-ed58effc1a09)
![PXL_20241004_120610965](https://github.com/user-attachments/assets/1107ecea-bdb7-4538-907b-262bbd0bf7bc)
![PXL_20241004_120620564](https://github.com/user-attachments/assets/56cfcb33-35d4-4249-9457-bfbe091d961a)
![PXL_20241004_120531947 MP](https://github.com/user-attachments/assets/501a7e80-85f2-4da0-b90c-4c2e7233e137)
![PXL_20241004_120539951](https://github.com/user-attachments/assets/8d1979c3-2bda-4f1e-bebc-4ce26cb212c4)
![PXL_20241004_120641826](https://github.com/user-attachments/assets/c24c4c78-c500-4ea8-ae49-6ab44e991821)
![PXL_20241004_120544606](https://github.com/user-attachments/assets/99f48471-3223-4f16-936e-0cfeddc3d8be)

**Comment by kachurovskiy** [comment]
Hi @leehambley, thanks for sharing with us! Hope the little H2 is helpful in your work.

**Comment by andreev-** [comment]
Hello, please tell me if you turn the spindle for some time, will the degree readings be displayed correctly?

**Comment by leehambley** [comment]
I don't lose synchronisation here, but there is a reduction between the encoder and the spindle. At higher rpms it does struggle and I think the Stepper is not strong enough

**Reply by andreev-** [reply]
https://github.com/user-attachments/assets/5d36752e-bd24-48ed-a669-436c502f9101

**Reply by andreev-** [reply]
Why can this happen?

**Reply by kachurovskiy** [reply]
Try changing https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino#L51 `DIR_CHANGE_DIFF_MICROS` to 0:

```
#define DIR_CHANGE_DIFF_MICROS (int) 0
```

This disables the steady-direction logic and always reads encoder direction from the pin. Sometimes disabling it helps with angle problems.

**Reply by andreev-** [reply]
doesn't help

**Comment by leehambley** [comment]
@kachurovskiy I wanted to check if you still had the driver config files for one of your builds around?

When I ran the geared setup above the ELS kind of worked, with timing pulleys it's faulting out due to overload every time I use it. I switched to pulleys because the gears were noisy. I'm also on a 50v PSU with 6A limit now (vs 24V@2A ).

I've tried I think every combination of stiffness, inertia, gain control and everything else, but regardless, I essentially cannot run the machine at all anymore, I've tried microsteps from 4000 to 400 and nothing I do works.

I may have overlooked in the repository if there was a motor parameters file, but I at least didn't find one. I'm now even running a 2:1 reduction to get more torque to the lead screw.

Do you have any advice or suggestions before I convert the machine back to mechanical control ?

![image](https://github.com/user-attachments/assets/737a9879-4cc3-443c-9600-bb5312f2caa9)

**Reply by kachurovskiy** [reply]
> driver config files for one of your builds

Not sure what you mean. The main file for the Arduino is https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino

From what I see on the picture above, the problems are: 

1. Switch the pulleys - large on the lead screw, small needs to be on the motor
2. Ensure you have low resistance on the motor wires - currently I see very thin signal wires with poor contact
3. Post more pictures covering your entire setup?
4. What's your model of the stepper and the driver?
5. What is the voltage you're running them on?

Thanks.

**Comment by leehambley** [comment]
Thanks @kachurovskiy <3 

1. Thanks. Not sure how I screwed that up, I was convinced I had a 2:1 _reduction_ and ChatGPT helpfully lied to to me repeatedly!
2. The signal wires are normal "jumper wire" (sparkfun says 30AWG) but I can beef them up for sure.
3. This was my build thread here, but I can post more pictures later.
4. My big mistake, I think. I started with the Cloud42 ELS but couldn't source the TI controller he used, so it's a StepperOnline Nema23 0.6nM closed-loop "servo" stepper. ([link](https://www.omc-stepperonline.com/de/nema-23-integrierter-easy-servo-motor-180w-3000rpm-0-6nm-84-98oz-in-20-50vdc-servomotor-kurze-welle-isv57t-180s)) the torque is really too weak though I think. 0.6nM, then I've gone and reduced it by half with my silly pulley setup 🤦 
5. Running a "48v" (actually reads 55) 6A MeanWell DIN rail PSU, so I trust this one, at least. Although it's weird that even with the trim screw all-hte-way down it is still 55v. The motor will take 20-50V. And has an integrated controller.

I'm tempted to move away from the closed-loop stepper, to a regular Nema23 with 2 or more nM of torque and that should be fine, no wonder I'm struggling with effectively 0.3nM of torque at my disposal now!

**Reply by kachurovskiy** [reply]
> The signal wires are normal "jumper wire" (sparkfun says 30AWG) but I can beef them up for sure.

These are signal wires, unsuitable for any meaningful power loads, you need at least 20 AWG / 0.5mm2 or better - ideally without any cuts in the middle between motor and power source.

> StepperOnline Nema23 0.6nM closed-loop "servo" stepper

That's an interesting motor but I would not recommend such low toque and placing the driver on the motor is fundamentally problematic since heat cycles and vibration kills electronics pretty fast.

I've settled on https://www.omc-stepperonline.com/ts-series-3-0-nm-424-92oz-in-1-axis-closed-loop-stepper-cnc-kit-nema-23-motor-driver-1-cl57t-s30a-v41 which have never let me down. Tried a lot of other motors and drivers. Just don't forget to switch the 24V signal toggle to 5V on the driver.

> trim screw all-the-way down it is still 55v

Perhaps try turning it all the way in the other direction? Over-voltage is not good though you probably loose some of those volts in the wires (but only when the motor is moving).

**Comment by leehambley** [comment]
The motor you linked is lovely indeed. Thanks for the suggestions. I'll update in a couple of weeks once I get it all working :) New pulleys are in the 3d printer in PLA-CF already to see how far I can get until the new motor and controller arrive. Happy holidays Maxim. LG aus Hamburg.

**Comment by leehambley** [comment]
![image](https://github.com/user-attachments/assets/61944198-97fd-45b5-9f92-ff564e202e0d)

With the pulleys the right way round, and 200 steps per rev, it works more than well enough to run a scratch pass on aluminum and I I can't stop the carriage with my fingers anymore!

**Comment by GeoGlobber** [comment]
I'm a big fan of the NanoEls project and currently working on implementing it myself. Wanted to clear something up about the two motors that are discussed here, for future readers like me. The servo that @leehambley uses (the 0.6N servo), if utilized right, has the potential to be (much!) more powerful than the stepper that @kachurovskiy recommends. It just needs additional reduction. Let me show you by comparing the torque curves (as publishing on the stepperonline website).

First is the curve for the 3N stepper:
![image](https://github.com/user-attachments/assets/4936e7e3-0e45-419f-a658-a98c8d61b184)
For the NanoEls, can we assume that a maximum RPM of 450-500 is enough? If so, we can use that to illustrate the point. As can be seen on this curve, the stepper motor has around 1.6Nm of torque around 450 RPM.

Here is the curve for the 0.6Nm servo:
![image](https://github.com/user-attachments/assets/f0fa5ceb-1857-4a30-ace9-f93a741ffc08)
The curve looks very different because the servo holds its torque all the way up to 3K RPM. That means you can use a 6:1 reduction (as Clough42 uses). With this reduction, the torque at 500 RPM is 3.6Nm (it wont be exactly but you get the point). What's more, this servo is able to hold a peak torque for around 10 seconds (according to stepperonline, although Clough42 has tested it to hold a peak torque for 23 seconds). Now imagine this peak torque, also multiplied.

I am not an expert by any means (and might be wrong), but I just wanted to point out that Clough42 has done some amazing research into these motors, and there is a reason why he ended up using this one. Hope this helps others that are looking for a motor to drive the NanoEls :smiley: 

Side question for @kachurovskiy , in this discussion you mention that having driver on the motor is a bad idea. In the hardware section I did see a recommendation for such a motor with integrated driver. Did you have some experience that caused you to change? Also, thank you very much for all the effort you put into this software!

**Reply by kachurovskiy** [reply]
Good points, @GeoGlobber. IIRC Clough42 also reported that the integrated motor died after 1 year. I've had people report some bad experiences with integrated motors but reviews on https://www.aliexpress.com/item/1005005598264428.html are surprisingly good so perhaps it's all clear now.

**Reply by leehambley** [reply]
Also, with a 6:1 reduction that would, I think be beyond the Arduino's step generating ability without a hardware timer. That would mean not being able to run the spindle faster than 140 RPM (mine won't go under 300).

Note in the h2 README [here](https://github.com/kachurovskiy/nanoels/blob/8cc4b13d93870869a071ace68bb6aeda617e1e79/h2/README.md#L27).

![image](https://github.com/user-attachments/assets/9b430cb8-8fd2-427e-a98a-236c79ad38f7)

(my motor runs at 200 pulses per rev, currently)

![image](https://github.com/user-attachments/assets/1ab559f4-eb15-477d-9a64-d2a5b16e98e7)


I think, honestly, it should be pretty trivial to port the H2 code to a faster MCU with better GPIO which is probably what I'll do next. 

I have for example a bunch of these laying around https://www.st.com/en/evaluation-tools/nucleo-l432kc.html#overview which are more or less the same size as the Nano.

The Nucleo has hardware timer support for reading from a quadrature encoder, and other timers (it has 3) which can be used for PWM output which makes driving the stepper pulses pretty easy.

**Reply by leehambley** [reply]
With a little digging in the ATmega328P docs, I think it should be possible to offload a lot of the work that H2 is doing currently in ISRs and with bit banging into hardware timers. I think I could conservatively get 10x more (i.e 25-30kHz) motor pulse resolution by doing more with hardware timers for the encoder inputs and motor outputs.

I'll dig deeper into the H2 code maybe next week. I know that @kachurovskiy is on to greener pastures now, but h4/h5 hold for me, currently no interest as I only want a "digital gearbox".

Maybe I can put 20 years software engineering into good practices and contribute some improved firmware for h2.

First best guesses:

* **Reducing or eliminating the pulse-by-pulse ISR:** (I think it can be eliminated see below)
* **Lower timer the prescalar:** If there's less code in the ISR, or no ISR we can use a higher frequency prescalar. (from 2Mhz to 16mhz?)
* **Stop using FastGPIO:** Fast GPIO is fast, but it's not fast enough. I think the pulse generation can be moved to a hardware timer which can automatically drive the PWM signal.

I think offloading both the encoder and stepper to hardware runs a risk of "drift", i.e the stepper might run at a flat "10khz" for 800ms or something, but maybe that's 3, 10, or 20k steps too many for the encoder input (maybe the spindle slowed under cutting load?). I _assume_ that with the encoder and stepper both running at the full 16Mhz of the ATmega328 that such drift would be tiny, but I guess along side the hardware offload, I'll need to use one of the 16 bit timers to count in and out the pulses atomically. I guess at 16 bits, that gives unsigned 65,535, so approximately 2-5 seconds depending how fast the spindle is running and how many steps per sec the motor is asked to run at.

(thinking aloud here and leaving myself some notes for next week)

**Reply by kachurovskiy** [reply]
You can find timer-based ELS implementation in https://github.com/kachurovskiy/megaels and tests of accumulating errors in https://youtu.be/AwgltElphHg?si=KmYdkxXO3AOofMoQ&t=1053
